File: app/src/main/java/com/niiadotei/storehouse/ui/stock/StockViewModel.java
package com.niiadotei.storehouse.ui.stock;


import android.annotation.SuppressLint;
import android.app.AlertDialog;
import android.app.Dialog;
import android.content.res.Resources;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Filter;
import android.widget.Filterable;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.RecyclerView;

import com.niiadotei.storehouse.R;
import com.niiadotei.storehouse.data.DatabaseHelper;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.text.DecimalFormat;
import java.text.NumberFormat;
import java.text.SimpleDateFormat;
import java.util.Arrays;
import java.util.Date;
import java.util.Locale;

public class StockViewModel extends RecyclerView.Adapter<StockViewModel.ViewHolder> implements Filterable {
    Fragment fragment;

    JSONArray jsonArray;
    JSONArray filteredJsonArray;

    DatabaseHelper databaseHelper;

    public FilteredViewHolder filteredViewHolder;


    public StockViewModel(Fragment fragment, JSONArray jsonArray) {
        this.fragment = fragment;
        this.jsonArray = jsonArray;

        filteredJsonArray = new JSONArray();
    }

    public void updateArray(JSONArray jsonArray) {
        this.jsonArray = jsonArray;
    }

    @NonNull
    @Override
    public StockViewModel.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_product, parent, false);

        databaseHelper = new DatabaseHelper(view.getContext());

        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull StockViewModel.ViewHolder holder, int position) {
        double price = 0.0;
        int id = 0;

        try {
            JSONObject jsonObject = jsonArray.getJSONObject(position);

            id = jsonObject.getInt("id");

            holder.displayNameTextView.setText(jsonObject.getString("display"));
            holder.nameTextView.setText(jsonObject.getString("name"));

            Resources resources = holder.itemView.getContext().getResources();
            Locale locale = resources.getConfiguration().locale;
            NumberFormat currencyInstance = DecimalFormat.getCurrencyInstance(locale);

            price = jsonObject.getDouble("price");
            holder.priceTextView.setText(currencyInstance.format(price));

            holder.supplierTextView.setText(databaseHelper.getSupplierNameFromID(jsonObject.getString("supplier")));

            String quantity = "Qty: " + jsonObject.getString("quantity");
            holder.quantityTextView.setText(quantity);
        } catch (JSONException e) {
            e.printStackTrace();
        }

        double finalPrice = price;
        int finalId = id;
        holder.itemView.setOnClickListener(view -> {
            String[] customers = databaseHelper.getAllCustomersWithPhone();
            Arrays.sort(customers);
            final String[] selectedCustomer = new String[1];
            selectedCustomer[0] = "";
            int[] checkedCustomer = {-1};

            Dialog dialog = new Dialog(holder.itemView.getContext());
            dialog.getWindow().setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
            dialog.setContentView(R.layout.dialog_purchase);
            dialog.show();

            EditText editText = dialog.findViewById(R.id.purchase_quantity);

            Button selectCustomer = dialog.findViewById(R.id.selectCustomerButton);
            selectCustomer.setOnClickListener(view1 -> {
                AlertDialog.Builder alertDialog = new AlertDialog.Builder(holder.itemView.getContext());
                alertDialog.setTitle("Select the customer buying");
                alertDialog.setSingleChoiceItems(customers, checkedCustomer[0], (dialogInterface, i) -> {
                    checkedCustomer[0] = i;
                    selectedCustomer[0] = customers[i];

                    selectCustomer.setText(selectedCustomer[0].substring(0, 10));
                    dialogInterface.dismiss();
                });

                alertDialog.setNegativeButton("Cancel", (dialogInterface, i) -> {

                });

                AlertDialog customerAlertDialog = alertDialog.create();
                customerAlertDialog.show();
            });

            Button makePurchase = dialog.findViewById(R.id.makePurchaseButton);
            makePurchase.setOnClickListener(view1 -> {
                try {
                    int quantity = Integer.parseInt(editText.getText().toString().trim());
                    int stockQuantity = databaseHelper.getProductQuantityFromID(finalId);

                    if (quantity == 0) {
                        Toast.makeText(holder.itemView.getContext(), "Quantity cannot be 0", Toast.LENGTH_LONG).show();
                        dialog.dismiss();
                        return;
                    }

                    if (quantity > stockQuantity) {
                        Toast.makeText(holder.itemView.getContext(), "Stock quantity not enough to make purchase", Toast.LENGTH_LONG).show();
                        dialog.dismiss();
                        return;
                    }

                    AlertDialog.Builder builder = new AlertDialog.Builder(fragment.getActivity());
                    builder.setTitle("Confirm");

                    double totalPurchaseCost = finalPrice * quantity;

                    Resources resources = holder.itemView.getContext().getResources();
                    Locale locale = resources.getConfiguration().locale;
                    NumberFormat currencyInstance = DecimalFormat.getCurrencyInstance(locale);

                    builder.setMessage("Collect from the customer, a total cost of " + currencyInstance.format(totalPurchaseCost) + " before confirming.");

                    builder.setPositiveButton("Yes", ((dialogInterface, i) -> {
                        try {
                            JSONObject jsonObject = jsonArray.getJSONObject(holder.getAdapterPosition());

                            int productID = jsonObject.getInt("id");

                            databaseHelper.updateProductQuantity(productID, -quantity);

                            databaseHelper.insertPurchases(selectedCustomer[0], holder.nameTextView.getText().toString(), totalPurchaseCost, quantity, new SimpleDateFormat("dd-MM-yyyy", Locale.getDefault()).format(new Date()));

                            Toast.makeText(holder.itemView.getContext(), "Purchase successfully made", Toast.LENGTH_LONG).show();

                            dialog.dismiss();
                        } catch (JSONException e) {
                            Toast.makeText(holder.itemView.getContext(), "Error making the purchase", Toast.LENGTH_LONG).show();
                            dialog.dismiss();
                        }
                    }));

                    builder.setNegativeButton("No", ((dialogInterface, i) -> dialogInterface.dismiss()));
                    builder.show();
                } catch (NumberFormatException e) {
                    Toast.makeText(holder.itemView.getContext(), "Quantity format not supported", Toast.LENGTH_LONG).show();
                }
            });

        });

        holder.itemView.setOnLongClickListener(view -> {
            int getPosition = holder.getAdapterPosition();

            try {
                JSONObject jsonObject = jsonArray.getJSONObject(getPosition);

                String stringID = jsonObject.getString("id");

                AlertDialog.Builder builder = new AlertDialog.Builder(fragment.getActivity());
                builder.setTitle("Confirm");
                builder.setMessage("Are you sure you want to delete this product?");

                builder.setPositiveButton("Yes", ((dialogInterface, i) -> {
                    databaseHelper.deleteProduct(stringID);
                    jsonArray.remove(getPosition);
                    notifyItemRemoved(getPosition);
                    notifyItemRangeRemoved(getPosition, jsonArray.length());
                }));

                builder.setNegativeButton("No", ((dialogInterface, i) -> dialogInterface.dismiss()));
                builder.show();
            } catch (JSONException e) {
                e.printStackTrace();
            }

            return true;
        });
    }

    @Override
    public Filter getFilter() {
        if (filteredViewHolder == null) {
            filteredJsonArray = jsonArray;

            filteredViewHolder = new StockViewModel.FilteredViewHolder(this, filteredJsonArray);
        }
        return filteredViewHolder;
    }

    public JSONArray getFilteredJsonArray(String name) {
        return databaseHelper.getFilteredProductArray(name);
    }

    @Override
    public int getItemCount() {
        return jsonArray.length();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        TextView displayNameTextView;
        TextView nameTextView;
        TextView priceTextView;
        TextView supplierTextView;
        TextView quantityTextView;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);

            displayNameTextView = itemView.findViewById(R.id.displayProductName);
            nameTextView = itemView.findViewById(R.id.productName);
            priceTextView = itemView.findViewById(R.id.productPrice);
            supplierTextView = itemView.findViewById(R.id.productSupplier);
            quantityTextView = itemView.findViewById(R.id.productQuantity);
        }
    }

    private static class FilteredViewHolder extends Filter {
        private final StockViewModel stockViewModel;

        private final JSONArray jsonArray;
        private JSONArray filteredJsonArray;

        private FilteredViewHolder(StockViewModel stockViewModel, JSONArray jsonArray) {
            this.stockViewModel = stockViewModel;

            this.jsonArray = jsonArray;
            this.filteredJsonArray = new JSONArray();
        }

        @Override
        protected FilterResults performFiltering(CharSequence charSequence) {
            final FilterResults results = new FilterResults();

            if (charSequence.length() == 0) {
                filteredJsonArray = jsonArray;
            } else {
                final String filterPattern = charSequence.toString().trim();

                filteredJsonArray = stockViewModel.getFilteredJsonArray(filterPattern);
            }

            results.values = filteredJsonArray;
            return results;
        }

        @SuppressLint("NotifyDataSetChanged")
        @Override
        protected void publishResults(CharSequence charSequence, FilterResults filterResults) {
            stockViewModel.updateArray((JSONArray) filterResults.values);
            stockViewModel.notifyDataSetChanged();
        }
    }
}

File: app/src/main/java/com/niiadotei/storehouse/ui/stock/StockViewModel.kt
package com.niiadotei.storehouse.ui.stock

import org.json.JSONArray
import androidx.recyclerview.widget.RecyclerView
import com.niiadotei.storehouse.data.DatabaseHelper
import android.view.ViewGroup
import android.view.LayoutInflater
import com.niiadotei.storehouse.R
import org.json.JSONException
import android.graphics.drawable.ColorDrawable
import android.content.DialogInterface
import android.text.TextUtils
import android.annotation.SuppressLint
import android.app.AlertDialog
import android.app.Dialog
import android.graphics.Color
import android.view.View
import android.widget.*
import androidx.fragment.app.Fragment
import java.lang.NumberFormatException
import java.text.DecimalFormat
import java.text.SimpleDateFormat
import java.util.*

class StockViewModel(var fragment: Fragment, var jsonArray: JSONArray) : RecyclerView.Adapter<StockViewModel.ViewHolder>(), Filterable {
    private var filteredJsonArray: JSONArray
    private lateinit var databaseHelper: DatabaseHelper

    init {
        filteredJsonArray = JSONArray()
    }

    fun updateArray(jsonArray: JSONArray) {
        this.jsonArray = jsonArray
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_product, parent, false)
        databaseHelper = DatabaseHelper(view.context)

        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        var price = 0.0
        var id = 0

        try {
            val jsonObject = jsonArray.getJSONObject(position)

            id = jsonObject.getInt("id")

            holder.displayNameTextView.text = jsonObject.getString("display")
            holder.nameTextView.text = jsonObject.getString("name")

            val resources = holder.itemView.context.resources
            val locale = resources.configuration.locale
            val currencyInstance = DecimalFormat.getCurrencyInstance(locale)

            price = jsonObject.getDouble("price")
            holder.priceTextView.text = currencyInstance.format(price)

            holder.supplierTextView.text = databaseHelper.getSupplierNameFromID(jsonObject.getInt("supplier"))

            val quantity = "Qty: " + jsonObject.getString("quantity")
            holder.quantityTextView.text = quantity
        } catch (e: JSONException) {
            e.printStackTrace()
        }

        val finalPrice = price
        val finalId = id

        holder.itemView.setOnClickListener {
            val customers = databaseHelper.allCustomersWithPhone
            Arrays.sort(customers)

            val selectedCustomer = arrayOfNulls<String>(1)
            selectedCustomer[0] = ""

            val checkedCustomer = intArrayOf(-1)

            val dialog = Dialog(holder.itemView.context)
            dialog.window!!.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
            dialog.setContentView(R.layout.dialog_purchase)
            dialog.show()

            val editText = dialog.findViewById<EditText>(R.id.purchase_quantity)
            val selectCustomer = dialog.findViewById<Button>(R.id.selectCustomerButton)

            selectCustomer.setOnClickListener {
                val alertDialog = AlertDialog.Builder(holder.itemView.context)
                alertDialog.setTitle(R.string.select_customer_message)
                alertDialog.setSingleChoiceItems(customers, checkedCustomer[0]) { dialogInterface: DialogInterface, i: Int ->
                    checkedCustomer[0] = i
                    selectedCustomer[0] = customers[i]
                    selectCustomer.text = selectedCustomer[0]!!.substring(0, 10)
                    dialogInterface.dismiss()
                }

                alertDialog.setNegativeButton(R.string.negative_cancel_button) { _: DialogInterface?, _: Int -> }
                val customerAlertDialog = alertDialog.create()
                customerAlertDialog.show()
            }

            val makePurchase = dialog.findViewById<Button>(R.id.makePurchaseButton)
            makePurchase.setOnClickListener MakePurchase@{
                try {
                    val quantity = editText.text.toString().trim { it <= ' ' }.toInt()
                    val stockQuantity = databaseHelper.getProductQuantityFromID(finalId)
                    if (quantity == 0) {
                        Toast.makeText(holder.itemView.context, R.string.zero_quantity_message, Toast.LENGTH_LONG).show()
                        dialog.dismiss()

                        return@MakePurchase
                    }
                    if (quantity > stockQuantity) {
                        Toast.makeText(holder.itemView.context, R.string.quantity_not_enough_message, Toast.LENGTH_LONG).show()
                        dialog.dismiss()

                        return@MakePurchase
                    }
                    if (TextUtils.isEmpty(selectedCustomer[0])) {
                        Toast.makeText(holder.itemView.context, R.string.empty_customer_message, Toast.LENGTH_LONG).show()

                        return@MakePurchase
                    }
                    val builder = AlertDialog.Builder(fragment.activity)
                    builder.setTitle(R.string.confirm_title)
                    val totalPurchaseCost = finalPrice * quantity

                    val resources = holder.itemView.context.resources
                    val locale = resources.configuration.locale
                    val currencyInstance = DecimalFormat.getCurrencyInstance(locale)

                    builder.setMessage("Collect from the customer, a total cost of " + currencyInstance.format(totalPurchaseCost) + " before confirming.")
                    builder.setPositiveButton(R.string.positive_button) { _: DialogInterface?, _: Int ->
                        try {
                            val jsonObject = jsonArray.getJSONObject(holder.adapterPosition)

                            val productID = jsonObject.getInt("id")

                            databaseHelper.updateProductQuantity(productID, -quantity)

                            databaseHelper.insertPurchases(selectedCustomer[0], holder.nameTextView.text.toString(), totalPurchaseCost, quantity, SimpleDateFormat("dd-MM-yyyy", Locale.getDefault()).format(Date()))

                            Toast.makeText(holder.itemView.context, R.string.purchase_success_message, Toast.LENGTH_LONG).show()
                            dialog.dismiss()
                        } catch (e: JSONException) {
                            Toast.makeText(holder.itemView.context, R.string.purchase_error_message, Toast.LENGTH_LONG).show()
                            dialog.dismiss()
                        }
                    }

                    builder.setNegativeButton(R.string.negative_button) { dialogInterface: DialogInterface, _: Int -> dialogInterface.dismiss() }
                    builder.show()
                } catch (e: NumberFormatException) {

                }
            }
        }

        holder.itemView.setOnLongClickListener {
            val getPosition = holder.adapterPosition
            try {
                val jsonObject = jsonArray.getJSONObject(getPosition)

                val stringID = jsonObject.getString("id")

                val builder = AlertDialog.Builder(fragment.activity)
                builder.setTitle(R.string.confirm_title)
                builder.setMessage(R.string.delete_product_message)
                builder.setPositiveButton(R.string.positive_button) { _: DialogInterface?, _: Int ->
                    databaseHelper.deleteProduct(stringID)

                    jsonArray.remove(getPosition)

                    notifyItemRemoved(getPosition)
                    notifyItemRangeRemoved(getPosition, jsonArray.length())
                }
                builder.setNegativeButton(R.string.negative_button) { dialogInterface: DialogInterface, _: Int -> dialogInterface.dismiss() }
                builder.show()
            } catch (e: JSONException) {
                e.printStackTrace()
            }
            true
        }
    }

    override fun getFilter(): Filter {
        return object: Filter() {
            override fun performFiltering(charSequence: CharSequence): FilterResults {
                val results = FilterResults()

                filteredJsonArray = if (charSequence.isEmpty()) {
                    jsonArray
                } else {
                    val filterPattern = charSequence.toString().trim { it <= ' ' }
                    getFilteredJsonArray(filterPattern)
                }

                results.values = filteredJsonArray
                return results
            }

            @SuppressLint("NotifyDataSetChanged")
            override fun publishResults(charSequence: CharSequence, filterResults: FilterResults) {

                if (charSequence.isEmpty()) {
                    updateArray(jsonArray)
                } else {
                    updateArray(filterResults.values as JSONArray)

                }

                notifyDataSetChanged()
            }
        }
    }

    fun getFilteredJsonArray(name: String?): JSONArray {
        return databaseHelper.getFilteredProductArray(name!!)
    }

    override fun getItemCount(): Int {
        return jsonArray.length()
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var displayNameTextView: TextView
        var nameTextView: TextView
        var priceTextView: TextView
        var supplierTextView: TextView
        var quantityTextView: TextView

        init {
            displayNameTextView = itemView.findViewById(R.id.displayProductName)
            nameTextView = itemView.findViewById(R.id.productName)
            priceTextView = itemView.findViewById(R.id.productPrice)
            supplierTextView = itemView.findViewById(R.id.productSupplier)
            quantityTextView = itemView.findViewById(R.id.productQuantity)
        }
    }


}